src/playback-watcher.js

```diff
@@ -107,18 +118,71 @@ export default class PlaybackWatcher {
     const buffered = this.tech_.buffered();
+    const atEndBuffer = !buffered.length ||
+      currentTime + Ranges.SAFE_TIME_DELTA >= buffered.end(buffered.length - 1);
+    // If current time is at the end of the final buffered region, then any playback
+    // stall is most likely caused by buffering in a low bandwidth environment. The tech
+    // should fire a `waiting` event in this scenario, but due to browser and tech
+    // inconsistencies. Calling `techWaiting_` here allows us to simulate
+    // responding to a native `waiting` event when the tech fails to emit one.
+
+    if (this.lastRecordedTime === currentTime && atEndBuffer) {
+      if (this.techWaiting_()) {
+        return;
+      }
+
+      const previousBytes = this.stallStartBytes_;
+
+      this.stallStartBytes_ = this.vhs.stats.mediaBytesTransferred;
+
+      if (previousBytes && previousBytes !== this.stallStartBytes_) {
+        this.stallByteChanges_++;
+      }
+
+      const mpc = this.vhs.masterPlaylistController_;
+
+      // if we have yet to be stalled for 3 segment downloads
+      // or mpc has an error, do nothing
+      if (this.stallByteChanges_ < 3 || mpc.error) {
+        return;
+      }
 
-    if (this.lastRecordedTime === currentTime &&
-        (!buffered.length ||
-         currentTime + Ranges.SAFE_TIME_DELTA >= buffered.end(buffered.length - 1))) {
-      // If current time is at the end of the final buffered region, then any playback
-      // stall is most likely caused by buffering in a low bandwidth environment. The tech
-      // should fire a `waiting` event in this scenario, but due to browser and tech
-      // inconsistencies. Calling `techWaiting_` here allows us to simulate
-      // responding to a native `waiting` event when the tech fails to emit one.
-      return this.techWaiting_();
+      this.resetStallHandlers_();
+      this.tech_.trigger({type: 'usage', name: 'hls-infinite-download-blacklist'});
+      const AUDIO = mpc.mediaTypes_.AUDIO;
+
+      // try blacklisting audio first
+      if (AUDIO.activePlaylistLoader && Object.keys(AUDIO.tracks).length > 1) {
+        const availableTracksIds = Object.keys(AUDIO.tracks)
+          .filter((id) => !AUDIO.tracks[id].enabled && !AUDIO.tracks[id].blacklist);
+
+        if (availableTracksIds.length) {
+          const activeTrack = AUDIO.activeTrack();
+          const media = AUDIO.activePlaylistLoader.media();
+
+          activeTrack.blacklist = true;
+          media.excludeUntil = Infinity;
+
+          AUDIO.tracks[availableTracksIds[0]].enabled = true;
+          this.tech_.audioTracks().removeTrack(activeTrack);
+          videojs.log.warn(`Problem encountered with playlist ${media.id}. ` +
+            `Playback stalled on audio ${activeTrack.id}. Switching to ${availableTracksIds[0]}.`);
+          return;
+        }
+      }
+
+      // if that didn't work or can't happen blacklist main media
+      const playlist = mpc.media();
```

Comment:
If we cannot blacklist any more audio playlists, and playback is still stalled try to blacklist main media playlists.